zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2216,4 +2253,25 @@ private void validateCurrentThread() {
       throw new IllegalArgumentException("Must not be done in the zookeeper event thread.");
     }
   }
+
+  private void checkNumChildren(String path) throws KeeperException, InterruptedException {
+    Stat stat = ((ZkConnection) getConnection()).getZookeeper().exists(path, false);
```

Comment:
I just noticed this issue. So you only check on connection loss exception. What if when you call this exists, the connection has not been recovered yet? This is very possible since there is no other operation or waits between the error and this exists() call.

In this case, the check will throw an exception, and the retry until connect call will also return unexpectedly.

I think we need to wait until connected here.

Reply:
@jiajunwang Good catch! I used to call `getStat()` considering reconnect. By somehow I changed it to `exists()` which is not supposed to. I believe `getStat()` is what we need: wait until connected and monitoring. Regarding nested retry until connected, it is fine because if getStat keeps getting connection loss and retrying, even we don't use getStat(), it'll be retrying as well in getChildren. 
Reverted to `getStat()`